### ISSUE TITLE
Unexpected information in top/bottom levels

### ISSUE THREAD

**Issue by pabloppp** [rosinality/vq-vae-2-pytorch#15]
I have done a custom PyTorch implementation of the vq-vae-2 using a very similar approach to what you do in this code, and I manage to get pretty accurate regenerated images, but the problem comes when looking at what the output of the decoder is, using only top or bottom encodings. 

The original paper says that top-level is supposed to encode the basic global information the image, while the bottom should only learn to encode details. They even show this picture depicting how, only by using the top-level the results are already recognizable, and middle/bottom layers only add detail information.

<img width="617" alt="Captura de pantalla 2019-08-25 a las 12 28 16" src="https://user-images.githubusercontent.com/4237014/63648708-f178dc00-c733-11e9-9b7e-e76f5380a35e.png">

But when I try to do the same and display the decoded image using only top and bottom encodings I get this:

![decomposed](https://user-images.githubusercontent.com/4237014/63648734-40267600-c734-11e9-8188-a4a5c45ca314.png)

As you can see, using both top+bottom gives a pretty decent reconstruction, but contrary to the expected, my bottom level seems to be encoding the main bulk of information about lines, shapes, and colors, instead of just extra details while top-level only encode color correction information.

This, of course, makes it really hard to generate new samples, because even if I was able to generate decent reconstructions of the top-level encodings, the information there is very scarce and the model in charge of generating new bottom-level encodings conditioned with the top-level ones has very little information to work with.

Is this something expected? Is my implementation missing something? What are the outputs of decoding using the top level in your implementation? I don't see anything in the original paper preventing this to happen.

Thank you, I appreciate your time for reading this!

PD: In case you may think that the problem is related to the data, let me show you a previous test using CIFAR10
![0748e393-69ae-46ed-89fd-c61027f21875](https://user-images.githubusercontent.com/4237014/63648846-e9ba3700-c735-11e9-9f4f-717cf4d38fa0.jpeg)
In this case, for some reason, the top-level learned to only encode colors while the bottom-level was encoding the black & white representation, not exactly the same as what I show above but still not what they show in the original paper.

**Comment by rosinality** [comment]
Hello, sorry for late reply. I don't know how authors did this experiments, so I simply substitutes each codes to ones at each levels.

![ffhq](https://user-images.githubusercontent.com/4343568/63993062-c7eff400-cb29-11e9-97e1-b3aa23038b36.png)

Top/Bottom/Both examples. I think it works correctly. I don't know the reason of these differences, but as this learned feature hierarchy is arbitrary, I think maybe training procedures or subtle differences in datasets can affect much to the results.

**Comment by pabloppp** [comment]
Thank you very much! your results look much more like they should, based on the paper!
I'll review my implementation to see if something's wrong with it, and if not I guess the difference is probably related to the data.

Thank you!

**Comment by SURABHI-GUPTA** [comment]
@rosinality how did you get the above top/bottom/both examples. please help. I tried but bottom level almost black image while top shows something about image.

**Comment by rosinality** [comment]
@SURABHI-GUPTA You can simply try

```python
quant_t, quant_b, diff, _, _ = vqvae.encode(img_t)
recon = vqvae.decode(quant_t * 0, quant_b).detach()[0].squeeze(0).add(1).div(2)
```

for the cases that only uses bottom latent codes for reconstruction, or

```python
quant_t, quant_b, diff, _, _ = vqvae.encode(img_t)
recon = vqvae.decode(quant_t, quant_b * 0).detach()[0].squeeze(0).add(1).div(2)
```

for reconstruction using only top codes.

**Comment by SURABHI-GUPTA** [comment]
@rosinality yes thanks.... I solved that... using your pretrained model I get bottom as almost black image, but when I trained the network from scratch using my own dataset, it worked. Now I am getting bottom/topbut not as you have shown above.


![decoded_sample_b](https://user-images.githubusercontent.com/22388014/95491361-1c190900-09b7-11eb-96cc-f12068ac3f94.png) ![decoded_sample_t](https://user-images.githubusercontent.com/22388014/95491371-21765380-09b7-11eb-876b-4d797866e0d3.png)

**Comment by SURABHI-GUPTA** [comment]
@rosinality can you help me with with the top/bottom reconstructions. by changing the number of embedding from 512 to 128, I got this output. I am really not able to figure out the detials that is being learned for top and bottom
![1](https://user-images.githubusercontent.com/22388014/98651755-a959cf80-2360-11eb-8ac5-ba298a31b7a2.png)

**Comment by rosinality** [comment]
@SURABHI-GUPTA Seems like that most informations are encoded in the bottom levels and top level only provides some color informations. Maybe it can be somewhat unexpected results, but it could be happen as there are no hard constraints to make top/bottom levels to learn to encode specific type of informations.

**Comment by SURABHI-GUPTA** [comment]
@rosinality I tried with FFHQ dataset, I think the top and bottom reconstructions are opposite of what it should be. According to what you have shown above with a girl picture, I am getting the opposite. The top should be global info and the bottom as local details. What can be the issue ?

![1](https://user-images.githubusercontent.com/22388014/101596001-54ba7a80-3a1a-11eb-8eb6-a051a2fb4e8a.png)

**Comment by rosinality** [comment]
@SURABHI-GUPTA Did you zeroed latent codes correctly? Maybe it is possible model to be trained like it.

**Comment by SURABHI-GUPTA** [comment]
> @SURABHI-GUPTA Did you zeroed latent codes correctly? Maybe it is possible model to be trained like it.

yes, I checked that thoroughly. I am not able to figure out where the problem is.

The code is like this,

    img = Image.open(img_path).convert('RGB')
    img_t = transform(img).unsqueeze(0)
    qt, qb, di, _id_t, _id_b = vqvae.encode(img_t)
    decoded_sample_bot = vqvae.decode(qt * 0, qb).detach()[0].squeeze(0)
    decoded_sample_top = vqvae.decode(qt, qb*0).detach()[0].squeeze(0)

**Comment by SURABHI-GUPTA** [comment]
@rosinality even using your pretrained model, I am getting reconstructions like this.

**Comment by rosinality** [comment]
I think you have flipped the results. I have tested it with my pretrained model and the result looks correct.

<img width="454" alt="test1" src="https://user-images.githubusercontent.com/4343568/102167693-406ef580-3ed2-11eb-9b7a-1a607b7f205e.PNG">

**Comment by Yoonho-Na** [comment]
this is strange that I also get all informations only in the bottom latents and top latents only show artifacts.
```
quant_t, quant_b, _, _, _ = vqvae.encode(input)
rec_img_only_b = vqvae.decode(quant_t*0, quant_b)
rec_img_only_t = vqvae.decode(quant_t, quant_b*0)
```

I just changed your code in 3D with single channel.
